actor/src/main/scala/org/apache/pekko/util/UniqueRandomShortProvider.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * license agreements; and to You under the Apache License, version 2.0:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * This file is part of the Apache Pekko project, derived from Akka.
+ */
+
+package org.apache.pekko.util
+
+import org.apache.pekko.annotation.InternalApi
+
+import java.security.SecureRandom
+
+/**
+ * An efficient collision free threadsafe generator of pseudo random
+ * shorts based on https://en.wikipedia.org/wiki/Double_hashing#Enhanced_double_hashing
+ * using SecureRandom as a seed.
+ *
+ * The avoidance of collisions is achieved by using a Short Array buffer
+ * that is prefilled with a dimension of 65536 (~ 131072 bytes in size per
+ * instance of [[UniqueRandomShortProvider]]). This means it will loop
+ * through all possible 65536 values before starting to re-generating
+ * the same numbers again.
+ */
+
+@InternalApi
+private[pekko] class UniqueRandomShortProvider {
+  import UniqueRandomShortProvider._
+
+  @volatile private var index = 0L
+  @volatile private var increment = 0L
+  @volatile private var count = 0L
+  @volatile private var limit = 0
+  private val buffer: Array[Short] = new Array[Short](UniqueRandomShortProvider.MaxLimit)
+  val r = new SecureRandom()
+
+  resetBuffer()
+  resetSeed()
+
+  count = 1
+
+  private def resetSeed(): Unit = {
+    index = r.nextLong
+    increment = r.nextLong
+  }
+
+  private def resetBuffer(): Unit = {
+    System.arraycopy(Filler, 0, buffer, 0, UniqueRandomShortProvider.MaxLimit)
```

Comment:
This seems not threadsafe

Reply:
I think we'll need to synchronize this block.

```
      resetBuffer()
      resetSeed()
```

Reply:
it's the recoverWith problem - the nextId() can be called by multiple threads - it seems like we need a synchronize somewhere

Reply:
The thread safety is a known problem and is being addressed separately, we discussed this earlier. I would recommend making the algorithm thread safe later, lets get an initial solution for now.

Using `synchronized` is usually just a heavy handed cop-out due to how slow it is, there are smarter ways to make it thread safe.

Reply:
@kerr we're working on an improvement to this solution and the thread safety issue will probably be easier to fix with this improvement. Let's get the improvement done and we can review if we can also fix that thread safety issue.

The adjustment is to @Claudenw's ShortSupplier and involves a change to stop us from potentially returning recently used IDs when we reach the resetBuffer stage.

Reply:
> The original code doesn't seem to worry about concurrency, and given that this PR just wants to introduce randomness, it's probably better to just drop those annotations — as they also introduced confusion, e.g., is this code supposed to be thread-safe or not; future readers will also wonder.

Definitely agreed, @Claudenw (who wrote the algoritm) did state that its not thread safe so we should just drop that `volatile`'s that he added there otherwise its giving false impressions.

However lets wait for https://github.com/apache/incubator-pekko/pull/371#issuecomment-1585680543

@alexandru Regarding scalafmt, its run on every CI run via the usage of scalafmt-native, see https://github.com/apache/incubator-pekko/actions/runs/5220862460/jobs/9424448652?pr=371 and https://github.com/apache/incubator-pekko/blob/main/.github/workflows/format.yml. Its actually not possible to merge a PR if it hasn't been formatted with scalafmt.

Reply:
@IainHull I think that the policies that you support in #385 are sufficient. You have a mechanism for tracking duplicate IDs in your PR. It's possible that there is stuff in this PR that may be a useful addition to the #385 changes but so far, I think they are rival solutions.

Reply:
@IainHull  Does it make sense to just make a new PR? Otherwise it seems to be a bit chaotic

Reply:
@mdedetrich I think the idea is that #385 could be merged instead of this. Is there anything that #385 is missing that we would need for a Pekko 1.0.0 release?